app/routines/create_user.rb

```diff
@@ -41,19 +29,23 @@ def exec(username:, title: nil, first_name: nil, last_name: nil,
     transfer_errors_from(outputs[:user], {type: :verbatim})
   end
 
-  # Returns a valid, though possibly non-unique, username
-  def get_valid_username(base)
-    base = randomify_username(base) if base.blank?
+  def generate_unique_valid_username(username)
+    return username if username_is_valid?(username) && User.new(username: username).username_is_unique?
 
-    # Go ahead and sanitize the username knowing that User.create will
-    # expect it to adhere to certain rules on length and content.
-    base = base.gsub(User::USERNAME_DISCARDED_CHAR_REGEX, '')
-               .slice(0..User::USERNAME_MAX_LENGTH - 1).downcase
+    username_max_attempts = 10
+
+    username_max_attempts.times do
+      username = User.new.create_random_username(7)
+      return username if User.new(username: username).username_is_unique?
+    end
+
+    raise "could not create a unique username after #{username_max_attempts} tries"
   end
 
-  def randomify_username(base)
-    base = 'user' if base.blank?
-    "#{base}#{rand(1000000)}"
+  def username_is_valid?(username)
+    user = User.new(username: username)
+    user.valid?
+    user.errors[:username].none?
   end
```

Comment:
Move this method into `User` as a static method `def self.username_is_valid?(username)`